exonum-java-binding-common/src/main/java/com/exonum/binding/common/proofs/list/ListProofStructureValidator.java

```diff
@@ -0,0 +1,182 @@
+ * limitations under the License.
+ */
+
+package com.exonum.binding.common.proofs.list;
+
+import com.exonum.binding.common.proofs.common.ProofStatus;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * A validator that checks list proofs internal structure.
+ */
+public final class ListProofStructureValidator implements ListProofVisitor {
+
+  public static int MAX_NODE_DEPTH = 64;
+
+  private List<NodeInfo> listProofBranchesInfo;
+
+  private List<NodeInfo> listProofElementsInfo;
+
+  private List<NodeInfo> listProofHashNodesInfo;
+
+  private int depth;
+
+  private ProofStatus proofStatus;
+
+  ListProofStructureValidator() {
+    depth = 0;
+    proofStatus = ListProofStatus.VALID;
+    listProofBranchesInfo = new ArrayList<>();
+    listProofElementsInfo = new ArrayList<>();
+    listProofHashNodesInfo = new ArrayList<>();
+  }
+
+  @Override
+  public void visit(ListProofBranch branch) {
+    int branchDepth = depth;
+
+    listProofBranchesInfo.add(new NodeInfo(branch, depth));
+    visitLeft(branch, branchDepth);
+    visitRight(branch, branchDepth);
+  }
+
+  @Override
+  public void visit(ListProofHashNode hashNode) {
+    listProofHashNodesInfo.add(new NodeInfo(hashNode, depth));
+  }
+
+  @Override
+  public void visit(ListProofElement element) {
+    listProofElementsInfo.add(new NodeInfo(element, depth));
+  }
+
+  private void visitLeft(ListProofBranch branch, int branchDepth) {
+    depth = getChildDepth(branchDepth);
+    branch.getLeft().accept(this);
+  }
+
+  private void visitRight(ListProofBranch branch, int branchDepth) {
+    depth = getChildDepth(branchDepth);
+    branch.getRight().ifPresent((right) -> right.accept(this));
+  }
+
+  private int getChildDepth(int branchDepth) {
+    return branchDepth + 1;
+  }
+
+  /**
+   * Returns list proof check status.
+   */
+  public ProofStatus check() {
+    if (exceedsMaxDepth(listProofElementsInfo)) {
+      proofStatus = ListProofStatus.INVALID_ELEMENT_NODE_DEPTH;
+      return proofStatus;
+    }
+
+    if (exceedsMaxDepth(listProofHashNodesInfo)) {
+      proofStatus = ListProofStatus.INVALID_HASH_NODE_DEPTH;
+      return proofStatus;
+    }
+
+    if (hasInvalidNodesDepth(listProofElementsInfo)) {
+      proofStatus = ListProofStatus.INVALID_NODE_DEPTH;
+      return proofStatus;
+    }
+
+    if (hasNoElementNodes(listProofBranchesInfo, listProofElementsInfo)) {
+      proofStatus = ListProofStatus.INVALID_TREE_NO_ELEMENTS;
+      return proofStatus;
+    }
+
+    return proofStatus;
+  }
+
+  /**
+   * Returns true if node exceeds the maximum depth at which nodes may appear.
+   *
+   * @param nodes collection of node info
+   * @return true if node depth is invalid
+   */
+  private boolean exceedsMaxDepth(List<NodeInfo> nodes) {
+    return nodes.stream()
+        .anyMatch(nodeInfo -> nodeInfo.getDepth() >= MAX_NODE_DEPTH);
```

Comment:
as I understand depth can be equal to max depth. No?

Reply:
We allow max depth as 63, maybe update it to 63 and nodeInfo.getDepth() > MAX_NODE_DEPTH, makes it more clear